chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,350 @@
+        m = F.cast(means, means.dtype)
+
+        module = cuda.get_array_module(means)
+        dd = module.asarray(delta_d, means.dtype)
+
+        dist_term = 0.0
+        for i in range(bs):
+
+            if n_objects[i] <= 1:
+                continue
+
+            nobj = n_objects[i]
+
+            # Prepare means
+            m_i = F.expand_dims(m[i, :nobj, :], 1)
+            m_1 = F.broadcast_to(m_i, (nobj, nobj, n_filters))
+            m_2 = F.transpose(m_1, axes=(1, 0, 2))
+
+            nrm = self.norm(m_1 - m_2, axis=2)
+            margin = 2.0 * dd * (1.0 - module.eye(nobj, dtype=means.dtype))
+
+            _dist_term_sample = F.sum(
+                F.maximum(module.asarray(0.0, means.dtype), margin - nrm) ** 2)
+            _dist_term_sample /= nobj * (nobj - 1)
+            dist_term += _dist_term_sample
+
+        dist_term /= bs
+
+        return dist_term
+
+    def _regularization_term(self, means, n_objects):
+        """Function to calculate regularization term
+
+        Args:
+            means (nd-array): Instance means
+            n_objects (int): Instance count in current input
+
+        Returns:
+            float : distance loss
+        """
+
+        bs, n_instances, n_filters = means.shape
+        m = F.cast(means, means.dtype)
+
+        reg_term = 0.0
+        for i in range(bs):
+            reg_term += F.mean(self.norm(m[i, : n_objects[i], :], 1))
+        reg_term /= bs
+
+        return reg_term
+
+    def _means(self, pred, gt, n_objects, max_n_objects, gt_idx):
+        """Function to calculate cluster means
+
+        Args:
+            pred (nd-array): Prediction output
+            gt (nd-array): Ground truth output
+            n_objects (int): Instance number in current input
+            max_n_objects (int): Maximum possible instance number
+            gt_idx (tuple / nd-array): Indexes of ground truth instances
+
+        Returns:
+            tuple : distance loss
+        """
+
+        bs = pred.shape[0]
+        n_filters = pred.shape[1]
+        n_instances = gt.shape[1]
+        n_loc = pred.shape[2]
+
+        # Expand prediction and broadcast instances to 3rd axis
+        p = F.reshape(pred, (bs, n_filters, 1, n_loc))
+        p = F.broadcast_to(p, (bs, n_filters, n_instances, n_loc))
+
+        # Expand ground truth to match the size but do not broadcast
+        g = F.reshape(gt, (bs, 1, n_instances, n_loc))
+
+        p = p * F.cast(g, p.dtype)
+        module = cuda.get_array_module(p)
+
+        means = []
+        for i in range(bs):
+
+            p_item = p[i, :, gt_idx[i]]
+            g_item = g[i, :, gt_idx[i]]
+
+            p_sum = F.sum(p_item, axis=2)
+            g_sum = F.cast(F.sum(g_item, axis=2), p_sum.dtype)
+            _mean_sample = p_sum / g_sum
+
+            n_fill_objects = max_n_objects - n_objects[i]
+
+            if n_fill_objects != 0:
+                _fill_sample = module.zeros(
+                    (module.asnumpy(n_fill_objects), n_filters),
```

Comment:
Do we have to convert `n_fill_objects` to NumPy array? I guess `module.zeros(n_fill_objects n_filters)` should work.

Reply:
Unfortunately, it does not work and we get `TypeError: 'cupy.core.core.ndarray' object cannot be interpreted as an integer` error. That's why I added this conversion. That's why I added this conversion.